### ISSUE TITLE
Decide on approach for reports

### ISSUE THREAD

**Issue by callsoft** [SharpsCompliance/PDSC#10]
I was thinking I would own a table in my portal database with report definitions (reportId, reportName, reportDescription, and parameter list/definitions).  I could then show this list to the user and upon selection by the user call your api with the reportId (or whatever key value you want to use) and the user selected parameters in the querystring and then you return the PDF that I would then provide to the user.

This would let us dynamically define reports as well as let you decided how the implementation should be done on your end without having to expose too much access to the front end.

**Comment by StuckerJ** [comment]
Yes.  This is fine.  Am already building similar for myself for some the mas 500 stock reports.  Will just expand for all.

If you put it in the API and have it communicate the report ID and parameters, that will work. Probably want to include the context cust Id (required) and addrid (optional) parameters.

I might expect to see 
/reports
contextCustID=12345
contextCustAddrID=12345
reportID=PackListSummary
Paramlist={“salesOrder”:”12345”}

Is that what you’re thinking?


The reports themselves -

Pack List
 parameter options that already exist–   
Shipment #
Ship Date
Sales Order #
Customer PO #

If you just want to pass the shipment # based on where they are navigating, that is OK.    That would simplify it, and it’s probably how it will be used.

Account Statement
-	Any cust addr ID sent will be ignored.  Invoices are only by Customer Main.  (We will have to decide whether location logins are even allowed to run a statement.)


Purchase History, active auto-ship, active auto-reorder, open invoice – these reports don’t already exist, so you tell me what parameter options you want and I’ll do it, assuming that if contextCustAddrId is sent, the results should be limited to having that address as a bill to or ship to on the record.

**Comment by callsoft** [comment]
The report parameters themselves would be up to you and Sharps, I'd just pass them through.

Yes I could see some parameters auto populating, but I'd also let them go to a general reports section in which case the user would have to select everything.

For the other reports, she indicated how she wanted that done with regard to parameters -> included below:

Pack List

    In the Orders tab, under individual Order Details, add button for option to print Pack List.
    Pull the Pack List as a PDF, as it currently pulls in MAS: Sales Order/Activities/Shipping/Print Pack List Summary. Example attached.

Account Statement

    Pull report as it pulls in MAS: Accounts Receivable/Insights/Report/Customer Statements. Example Attached.

For the following reports, attached is an excel spreadsheet showing examples of what each report will look like. Actual reports in portal need to pull as PDF copies.

 

 

Purchase History

    Pull report for one location or all locations, by specific item or all items.
    Set the default date range as 12 months but allow customer to select a custom date range as well.
    Group same items per location into one total QTY.

·         Transaction Date

·         Address Name

·         Item

·         Short Description

·         QTY

 

Active Auto-Ship Items

    Pull report for one location or all locations.
    Only if item does not have an end date.
    Exclude items that begin with “M” which is a monthly fee.

·         CustAddrID

·         AddrName

·         Item

·         ShortDesc

·         QTY

·         Frequency

·         Interval

 

Active Auto-Reorder Items

    If a QTY greater than (1) of the same item is purchased in the same sales order, list each item individually so customer can see individual serial number for each item.
    List only if SharpsTracer does not show the item as received.

·         Ship Date

·         AddrName

·         Item

·         Description

·         QTY

·         Manifest

 

Open Invoices

    Pull report for one location or all locations.

·         Invc Date

·         Invoice

·         Ship Addr Name

·         Invc Amt

·         Due
